src/drivers/samv7/tone_alarm/tone_alarm.cpp

```diff
@@ -512,384 +355,82 @@ ToneAlarm::stop_note()
 	/*
 	 * Make sure the GPIO is not driving the speaker.
 	 */
-	sam_configgpio(GPIO_TONE_ALARM_IDLE);
+	px4_arch_configgpio(GPIO_TONE_ALARM_IDLE);
 }
 
-void
-ToneAlarm::start_tune(const char *tune)
+void ToneAlarm::next_note()
 {
-	// kill any current playback
-	hrt_cancel(&_note_call);
-
-	// record the tune
-	_tune = tune;
-	_next = tune;
-
-	// initialise player state
-	_tempo = 120;
-	_note_length = 4;
-	_note_mode = MODE_NORMAL;
-	_octave = 4;
-	_silence_length = 0;
-	_repeat = false;		// otherwise command-line tunes repeat forever...
-
-	// schedule a callback to start playing
-	hrt_call_after(&_note_call, 0, (hrt_callout)next_trampoline, this);
-}
+	if (!_should_run) {
+		if (_tune_control_sub >= 0) {
+			orb_unsubscribe(_tune_control_sub);
+		}
 
-void
-ToneAlarm::next_note()
-{
-	// do we have an inter-note gap to wait for?
-	if (_silence_length > 0) {
-		stop_note();
-		hrt_call_after(&_note_call, (hrt_abstime)_silence_length, (hrt_callout)next_trampoline, this);
-		_silence_length = 0;
+		_running = false;
 		return;
 	}
 
-	// make sure we still have a tune - may be removed by the write / ioctl handler
-	if ((_next == nullptr) || (_tune == nullptr)) {
+	// subscribe to tune_control
+	if (_tune_control_sub < 0) {
+		_tune_control_sub = orb_subscribe(ORB_ID(tune_control));
+	}
+
+	// do we have an inter-note gap to wait for?
+	if (_silence_length > 0) {
 		stop_note();
+		work_queue(HPWORK, &_work, (worker_t)&ToneAlarm::next_trampoline, this, USEC2TICK(_silence_length));
```

Comment:
Smart, waiting for the next note using the scheduler :+1:

Reply:
So to add some important context that may be not obvious to the new user: A thread running in the HPWORK (or LPWORK) should not block. This means that sleep,usleep poll etc should not be used. So this is the only way to "sleep" with this execution pattern. 

The non blocking restriction is because all the threads of execution in the queue have the same priority and are serialized in that queue. Blocking one blocks all of them.